valohai_cli/utils/__init__.py

```diff
@@ -154,3 +155,25 @@ def parse_environment_variable_strings(
 
 def compact_dict(dct: dict) -> dict:
     return {key: value for (key, value) in dct.items() if key and value}
+
+
+T = TypeVar("T")
+
+
+def call_with_retry(
+    func: Callable[..., T],
+    args: Tuple = (),
+    kwargs: Optional[dict] = None,
+    retries: int = 3,
+    delay_range: Tuple[int, int] = (1, 5),
+    exception: Type[Exception] = Exception,
+) -> T:
+    kwargs = kwargs or {}
+    for attempt in range(retries):
+        try:
+            return func(*args, **kwargs)
+        except exception as e:
+            click.echo(f"{attempt + 1}/{retries} Retrying: {e}...")
+            time.sleep(random.randint(*delay_range))
+
+    raise exception(f"Failed after {retries} retries")
```

Comment:
* We don't need `args` and `kwargs` here; `func` itself could be a `lambda` or `functools.partial()`.
* We could want to catch multiple exception classes (e.g. `(IOError, TimeoutError)`)
* `raise exception` is a bit weird – you can't necessarily be certain that any given exception type passed in by the user is constructible with a string. It would also be useful to be able to access the exception.

Maybe something like:

```suggestion
T = TypeVar("T")


def call_with_retry(
    func: Callable[[], T],
    retries: int = 3,
    delay_range: Tuple[int, int] = (1, 5),
    retry_on_exception_classes: Tuple[Type[Exception], ...] = (Exception,),
) -> T:
    for attempt in range(retries):
        try:
            return func()
        except retry_on_exception_classes as e:
            if attempt == retries - 1:
                raise RuntimeError(f"Failed after {retries} attempts") from e
            click.echo(f"{attempt + 1}/{retries} Retrying: {e}...")
            time.sleep(random.uniform(*delay_range))
```